src/WebJobs.Script.WebHost/FileMonitoringService.cs

```diff
@@ -233,5 +238,44 @@ public void Dispose()
         {
             Dispose(true);
         }
+
+        private void OnOfflineFileChanged(FileSystemEventArgs e)
+        {
+            if (e.ChangeType == WatcherChangeTypes.Created)
+            {
+                // when app_offline.htm is created, trigger
+                // a shutdown so when the host starts back up it
+                // will be offline
+                Shutdown();
+            }
+            else if (e.ChangeType == WatcherChangeTypes.Deleted)
+            {
+                // after deleting app_offline.htm trigger a reinitialization
+                // to bring the host back online
+                _scriptEnvironment.RestartHost();
+            }
+        }
+
+        internal static void SetAppOfflineState(string rootPath, bool offline)
+        {
+            string path = Path.Combine(rootPath, ScriptConstants.AppOfflineFileName);
+            bool offlineFileExists = File.Exists(path);
+
+            if (offline && !offlineFileExists)
+            {
+                // create the app_offline.htm file in the root script directory
+                string content = FileUtility.ReadResourceString($"{ScriptConstants.ResourcePath}.{ScriptConstants.AppOfflineFileName}");
+                var bytes = Encoding.UTF8.GetBytes(content);
+                using (var fs = File.Create(path))
```

Comment:
Any chance this create call would race with the above exists call if we had two events that trigger offline state come in at close to the same time?

Reply:
Moved to use FileUtility